### ISSUE TITLE
hasOne relation not rollback when owner model is not valid and relation are inserted

### ISSUE THREAD

**Issue by douglasgsouza** [la-haute-societe/yii2-save-relations-behavior#45]
The hasOne relationships are saved and inserted in beforeValidate and if owner model not inserted due to validation error hasOne inserted relations not rollback.

How to Test:

In model Project, name attribute is required and if you don't set name, $project->save() will be failed, but company record willl be inserted in database.

public function testSaveNewHasOneRelationShouldSucceed()
    {
        $project = new Project();
        //$project->name = "Java";
        $company = new Company();
        $company->name = "Oracle";
        $project->company = $company;
        $this->assertTrue($company->isNewRecord, 'Company should be a new record');
        $this->assertTrue($project->save(), 'Project could not be saved');
        $this->assertNotNull($project->company_id, 'Company ID should be set');
        $this->assertEquals($project->company_id, $company->id, 'Company ID is not the one expected');
    }

**Comment by juban** [comment]
Did you activate the transaction on your owner model?
If not, that’s mandatory in order to correctly rollback related models.

**Comment by douglasgsouza** [comment]
> Did you activate the transaction on your owner model?
> If not, that’s mandatory in order to correctly rollback related models.

Yes. I used transactions() OP_ALL on owner model.

See more detailed:

Owner Model:
`<?php
namespace app\models;

use lhs\Yii2SaveRelationsBehavior\SaveRelationsBehavior;
use yii\behaviors\OptimisticLockBehavior;
use yii\db\ActiveRecord;

/**
 * This is the model class for table "sale".
 *
 * @property int $id
 * @property string $sale_date
 * @property string $contract
 * @property int $customer_id
 * @property int $version
 * @property Customer $customer
 * ... another properties
 */
class Sale extends ActiveRecord
{

    public static function tableName()
    {
        return 'sale';
    }

    public function rules()
    {
        return [
            [['sale_date', 'contract', 'customer_id'], 'required'],
            [['sale_date', 'customer'], 'safe'],
            [['customer_id', 'version'], 'integer'],
            [['contract'], 'string', 'max' => 20],
            [['status'], 'string', 'max' => 1],
            [['customer_id'], 'exist', 'skipOnError' => true, 'targetClass' => Customer::class, 'targetAttribute' => ['customer_id' => 'id']],
        ];
    }

    public function behaviors()
    {
        $behaviors = parent::behaviors();
        $behaviors['saveRelations'] = [
            'class'     => SaveRelationsBehavior::class,
            'relations' => [
                'customer',
            ],
        ];
        $behaviors['optimisticLock'] => [
            'class'             => OptimisticLockBehavior::class,
            'skipUpdateOnClean' => true
        ];
        return $behaviors;
    }

    public function optimisticLock()
    {
        return 'version';
    }

    public function getCustomer()
    {
        return $this->hasOne(Customer::class, ['id' => 'customer_id']);
    }

    public function transactions()
    {
        return [
            self::SCENARIO_DEFAULT => self::OP_ALL,
        ];
    }

}

`
Related Model:
`<?php

namespace app\models;
use yii\behaviors\OptimisticLockBehavior;
use yii\db\ActiveRecord;

/**
 * This is the model class for table "customer".
 *
 * @property int $id
 * @property string $name
 * @property string $document
 * @property int $version
 * ... another properties
 */

class Customer extends ActiveRecord
{
    public static function tableName()
    {
        return 'customer';
    }

    public function rules()
    {
        return [
            [['name', 'document'], 'required'],
            [['version'], 'integer'],
            [['name'], 'string', 'max' => 255],
            [['document'], 'string', 'max' => 14]
        ];
    }

    public function behaviors()
    {
        $behaviors = parent::behaviors();
        $behaviors['optimisticLock'] => [
            'class'             => OptimisticLockBehavior::class,
            'skipUpdateOnClean' => true
        ];
        return $behaviors;
    }

    public function optimisticLock()
    {
        return 'version';
    }

    public function transactions()
    {
        return [
            self::SCENARIO_DEFAULT => self::OP_ALL,
        ];
    }
}
`

Test:
`
$sale = new Sale();
$sale->sale_date = '2019-04-02';
//$sale->contract = '123'; //intentionally the mandatory property is not filled to cause validation error in the owner model

$sale->customer = ['name' => 'test customer', 'document' => '1234']; // Will create a new customer record

$sale->save();
`

Program Execution Steps:

$sale->save() -> call SaveRelationsBehavior
SaveRelationsBehavior->beforeValidate() ->
SaveRelationsBehavior->saveRelatedRecords() ->
SaveRelationsBehavior->_prepareHasOneRelation() -> related model save()
SaveRelationsBehavior->_prepareHasOneRelation() -> related model save() -> open Transaction
SaveRelationsBehavior->_prepareHasOneRelation() -> related model save() -> Insert Customer in DB
SaveRelationsBehavior->_prepareHasOneRelation() -> related model save() -> commit Transaction
$sale->save() -> continues and call owner model validate()
$sale->save() -> return false, Model not inserted due to validation error.

A validation error occurred in the owner model so the owner record was not entered in the database, but the related record was inserted into the database.

**Comment by douglasgsouza** [comment]
> Did you activate the transaction on your owner model?
> If not, that’s mandatory in order to correctly rollback related models.

Yes. I used transactions() OP_ALL on owner model.

See more detailed:

Owner Model:

```
<?php
namespace app\models;

use lhs\Yii2SaveRelationsBehavior\SaveRelationsBehavior;
use yii\behaviors\OptimisticLockBehavior;
use yii\db\ActiveRecord;

/**
 * This is the model class for table "sale".
 *
 * @property int $id
 * @property string $sale_date
 * @property string $contract
 * @property int $customer_id
 * @property int $version
 * @property Customer $customer
 * ... another properties
 */
class Sale extends ActiveRecord
{

    public static function tableName()
    {
        return 'sale';
    }

    public function rules()
    {
        return [
            [['sale_date', 'contract', 'customer_id'], 'required'],
            [['sale_date', 'customer'], 'safe'],
            [['customer_id', 'version'], 'integer'],
            [['contract'], 'string', 'max' => 20],
            [['status'], 'string', 'max' => 1],
            [['customer_id'], 'exist', 'skipOnError' => true, 'targetClass' => Customer::class, 'targetAttribute' => ['customer_id' => 'id']],
        ];
    }

    public function behaviors()
    {
        $behaviors = parent::behaviors();
        $behaviors['saveRelations'] = [
            'class'     => SaveRelationsBehavior::class,
            'relations' => [
                'customer',
            ],
        ];
        $behaviors['optimisticLock'] => [
            'class'             => OptimisticLockBehavior::class,
            'skipUpdateOnClean' => true
        ];
        return $behaviors;
    }

    public function optimisticLock()
    {
        return 'version';
    }

    public function getCustomer()
    {
        return $this->hasOne(Customer::class, ['id' => 'customer_id']);
    }

    public function transactions()
    {
        return [
            self::SCENARIO_DEFAULT => self::OP_ALL,
        ];
    }

}
```

Related Model:

```
<?php

namespace app\models;
use yii\behaviors\OptimisticLockBehavior;
use yii\db\ActiveRecord;

/**
 * This is the model class for table "customer".
 *
 * @property int $id
 * @property string $name
 * @property string $document
 * @property int $version
 * ... another properties
 */

class Customer extends ActiveRecord
{
    public static function tableName()
    {
        return 'customer';
    }

    public function rules()
    {
        return [
            [['name', 'document'], 'required'],
            [['version'], 'integer'],
            [['name'], 'string', 'max' => 255],
            [['document'], 'string', 'max' => 14]
        ];
    }

    public function behaviors()
    {
        $behaviors = parent::behaviors();
        $behaviors['optimisticLock'] => [
            'class'             => OptimisticLockBehavior::class,
            'skipUpdateOnClean' => true
        ];
        return $behaviors;
    }

    public function optimisticLock()
    {
        return 'version';
    }

    public function transactions()
    {
        return [
            self::SCENARIO_DEFAULT => self::OP_ALL,
        ];
    }
}
```

Test:


```
$sale = new Sale();
$sale->sale_date = '2019-04-02';
//$sale->contract = '123'; //intentionally the mandatory property is not filled to cause validation error in the owner model

$sale->customer = ['name' => 'test customer', 'document' => '1234']; // Will create a new customer record

$sale->save();
```

Program Execution Steps

$sale->save();
SaveRelationsBehavior->beforeValidate()
SaveRelationsBehavior->beforeValidate()->_prepareHasOneRelation()
SaveRelationsBehavior->beforeValidate()->_prepareHasOneRelation()-> call related save();
SaveRelationsBehavior->beforeValidate()->_prepareHasOneRelation()-> open Transaction
SaveRelationsBehavior->beforeValidate()->_prepareHasOneRelation()-> insert into DB
SaveRelationsBehavior->beforeValidate()->_prepareHasOneRelation()-> commit transaction
$sale->save(); -> continues and call model validate()
$sale->save(); -> return false, Model not inserted due to validation error.


A validation error occurred in the owner model, so the owner record was not entered into the database, but the child record was inserted into the database because it is inserted before and is not deleted when a validation error occurs in the owner model.

**Comment by juban** [comment]
Hi @douglasgsouza 
The issue should be fixed by 1.7.2 release.
Thank you for your patience and support.
